File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public List<String> solution(String[] record) {
        List<String> answer = new ArrayList<>();
        Map<String, String> dic = new HashMap<>();
        
        for (String r : record) {
            String[] data = r.split(" ");
            if (data[0].equals("Enter") || data[0].equals("Change"))
                dic.put(data[1], data[2]);
        }
        
        for (String r : record) {
            String[] data = r.split(" ");
            if (data[0].equals("Enter"))
                answer.add(dic.get(data[1]) + "님이 들어왔습니다.");
            else if (data[0].equals("Leave"))
                answer.add(dic.get(data[1]) + "님이 나갔습니다.");
        }
        
        return answer;
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    name = {}
    
    for r in record:
        r = r.split()
        action, uid = r[0], r[1]
        if action in ["Enter", "Change"]:
            name[uid] = r[2]
            
    for r in record:
        r = r.split()
        action, uid = r[0], r[1]
        if action == "Enter":
            answer.append(f"{name[uid]}님이 들어왔습니다.")
        elif action == "Leave":
            answer.append(f"{name[uid]}님이 나갔습니다.")
        
    return answer